### PR TITLE
Df.copy impl

### PR DESCRIPTION
**Pull request by 1e-to**

### CODE REVIEW

File: examples/dataframe/dataframe_copy.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import pandas as pd
+from numba import njit
+
+
+@njit
+def dataframe_copy():
+    df = pd.DataFrame({'A': [1.0, 2.0, 3.0, 1.0], 'B': [4, 5, 6, 7]})
+    new_df = df.copy(deep=True)
+    return new_df, df.values is new_df.values or df.index is new_df.index
```
Comment (densmirn) [inline examples/dataframe/dataframe_copy.py]: How I would check the `new_df` is really copy of `df`? Firstly do copy of `df` then change `df` and return `new_df`. If `new_df` is equal to `df`, сongratulations!

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 2]
```diff
@@ -556,6 +556,98 @@ def head_overload(df, n=5):
...
+            series_C = pandas.Series(get_dataframe_data(df, 2))
+            result_C = series_C.copy(deep=deep)
+            return pandas.DataFrame({"A": result_A, "B": result_B, "C": result_C}, index=df._index)
+    """
+    results = []
+    joined = ', '.join(func_params)
+    func_lines = [f'def _df_{func_name}_impl(df, {joined}):']
+    ind = df_index_codegen_all(df)
+    for i, c in enumerate(columns):
+        result_c = f'result_{c}'
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: Please take into account case when column name contains a white space.

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 3]
```diff
@@ -556,6 +556,98 @@ def head_overload(df, n=5):
...
+    data = ', '.join(f'"{col}": {data}' for col, data in results)
+    func_lines += [f'  return pandas.DataFrame({{{data}}}, {ind})']
+    func_text = '\n'.join(func_lines)
+    global_vars = {'pandas': pandas,
+                   'get_dataframe_data': get_dataframe_data}
+
+    return func_text, global_vars
+
+
+def sdc_pandas_dataframe_copy_codegen(df, func_name, params, ser_params):
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: Name `ser_params` looks a bit strange. Maybe rename it to `series_params`?

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 4]
```diff
@@ -556,6 +556,98 @@ def head_overload(df, n=5):
...
+
+    Limitations
+    -----------
+    - Parameter deep=False is currently unsupported for indexes by Intel Scalable Dataframe Compiler
+
+    Examples
+    --------
+    .. literalinclude:: ../../../examples/dataframe/dataframe_copy.py
+       :language: python
+       :lines: 35-
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: not 36?

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 5]
```diff
@@ -556,6 +556,94 @@ def head_overload(df, n=5):
...
+    Example func_text for func_name='copy' columns=('A', 'B', 'C'):
+        def _df_copy_impl(df, deep=True):
+            series_0 = pandas.Series(get_dataframe_data(df, 0))
+            result_0 = series_0.copy(deep=deep)
+            series_1 = pandas.Series(get_dataframe_data(df, 1))
+            result_1 = series_1.copy(deep=deep)
+            return pandas.DataFrame({"A": result_0, "B": result_1}, index=df._index)
+    """
+    results = []
+    joined = ', '.join(func_params)
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: `joined` -> `func_params_as_str` or `impl_params_as_str`

**Comment by PokhodenkoSA** [comment]
@1e-to could you write benchmark test too?

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 6]
```diff
@@ -556,6 +556,93 @@ def head_overload(df, n=5):
...
+    *************************************************
+    Pandas DataFrame method :meth:`pandas.Series.copy` implementation.
+
+    .. only:: developer
+        Test: python -m sdc.runtests -k sdc.tests.test_dataframe.TestDataFrame.test_df_copy*
+    """
+    ty_checker = TypeChecker("Method copy().")
+    ty_checker.check(df, DataFrameType)
+
+    if not isinstance(deep, (types.Omitted, types.Boolean)) and not deep:
```
Comment (kozlov-alexey) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: 'not deep' will evaluate to True for many things, not only False which we need to test here, so it's more proper to write, either:
```suggestion
    if not (isinstance(deep, (types.Omitted, types.Boolean)) or deep is True):
```
or equivalent:
```suggestion
    if not isinstance(deep, (types.Omitted, types.Boolean) and deep is not False:
```

File: sdc/tests/test_dataframe.py [hunk 7]
```diff
@@ -1006,6 +1006,23 @@ def impl(a):
                     )
                     pd.testing.assert_frame_equal(sdc_func(df), ref_impl(df))
 
+    def test_df_copy(self):
+        def test_impl(df, deep):
+            return df.copy(deep=deep)
+
+        sdc_func = sdc.jit(test_impl)
+        indexes = [[3, 4, 2, 6, 1], ['a', 'b', 'c', 'd', 'e'], None]
+        cases_deep = [None, True, False]
```
Comment (kozlov-alexey) [inline sdc/tests/test_dataframe.py]: Does None work? We never test for deep be of types.NoneType in the overload.

**Pull request merged**